### PR TITLE
Add `batched_generate_fn()`

### PR DESCRIPTION
**Pull request by apaz-cli**
Still a bit rough around the edges. Debug prints are commented out. But it does the thing, so putting it up for review and feedback. Writing tests and cleaning it up.

* Does handle:
  * Stop sequences (supports all litgpt models)
* Does not yet handle:
  * Batched `input_pos`
  * Padding to the left
  * `include_eos`

Once `include_eos` works, I'd like to rewrite `generate_fn()` in terms of `batched_generate_fn()`.

Also we're going to have to re-benchmark everything of course.

### CODE REVIEW

File: litgpt/generate/base.py [hunk 1]
```diff
@@ -194,6 +194,120 @@ def generate_fn(
...
+
+    if prompts.ndim == 1:
+        prompts = prompts.unsqueeze(0)
+
+    assert prompts.ndim == 2, "Prompts must be a 2D tensor."
+    batch_size = prompts.size(0)
+    max_prompt_size = prompts.size(1)
+    device = prompts.device
+
+    assert all(prompt.size(0) == max_prompt_size for prompt in prompts), "For now, prompts must have the same length."
```
Comment (t-vi) [inline litgpt/generate/base.py:222]: You specify and use that prompts is a tensor, then this does not make much sense.

File: litgpt/generate/base.py [hunk 2]
```diff
@@ -194,6 +194,120 @@ def generate_fn(
...
+    model: GPT,
+    prompts: torch.Tensor,
+    max_returned_tokens: int,
+    *,
+    sample_args: Union[list[dict], dict],
+    stop_tokens: Tuple[List[int], ...] = (),
+    include_prompt: bool,
+    include_eos: bool,
+) -> Iterator[list[Union[torch.Tensor, None]]]:
+
```
Comment (t-vi) [inline litgpt/generate/base.py:210]: Given the complexity of the this, would it make sense to have a docstring describing the args and what their relation needs to be?

File: litgpt/generate/base.py [hunk 3]
```diff
@@ -194,6 +194,120 @@ def generate_fn(
...
+    if prompts.ndim == 1:
+        prompts = prompts.unsqueeze(0)
+
+    assert prompts.ndim == 2, "Prompts must be a 2D tensor."
+    batch_size = prompts.size(0)
+    max_prompt_size = prompts.size(1)
+    device = prompts.device
+
+    assert all(prompt.size(0) == max_prompt_size for prompt in prompts), "For now, prompts must have the same length."
+    assert all(prompt.device == prompts[0].device for prompt in prompts), "Prompts must be on the same device."
```
Comment (t-vi) [inline litgpt/generate/base.py:223]: similar

File: litgpt/generate/base.py (hunk 2 above)
Comment (apaz-cli) [reply litgpt/generate/base.py:210]: Yeah, agreed. And also for `generate_fn()`.

The way that we've been thinking about it until this point is that, strictly speaking, anything that isn't accessible from a script or in `litgpt.api` isn't publicly facing. It's clear though that a lot of it is. We still have to go back and document a lot of that. Documenting these would be a good start.

File: litgpt/generate/base.py (hunk 1 above)
Comment (apaz-cli) [reply litgpt/generate/base.py:222]: The context around this changed, and it appears that I forgot to go back and edit it.

File: litgpt/generate/base.py (hunk 3 above)
Comment (apaz-cli) [reply litgpt/generate/base.py:223]: Same here

File: litgpt/generate/base.py [hunk 4]
```diff
@@ -194,6 +194,120 @@ def generate_fn(
...
+        tokens = batched_next_token(model, input_pos, tokens, sample_args)
+        for i in range(batch_size):
+            token_lists[i].append(tokens[i])
+        int_tokens = [token.item() for token in tokens]
+
+        # Check for stop sequences
+        # For each stop sequence, we keep a running total of how many are matched in stop_progress.
+        # If the current token matches the next token in the stop sequence, we increment the
+        # running total and hold off on yielding the token.
+        for batch_idx, int_token in enumerate(int_tokens):
```
Comment (t-vi) [inline litgpt/generate/base.py:260]: That we're basically doing this per batch-index makes me wonder what was the advantage of switching it to this loop from the one-line any condition.

File: litgpt/generate/base.py (hunk 2 above)
Comment (apaz-cli) [reply litgpt/generate/base.py]: Done

**Pull request merged**